### ISSUE TITLE
Action definition responses not enforced

### ISSUE THREAD

**Issue by frodebjerke** [goadesign/goa#836]
To my understanding it seems default handling of validation errors when validating payloads is not reflected in swagger by default. Therefore, if it is forgotten (likely) to add `Response(BadRequest, ErrorMedia)` the API design and API documentation will be out of sync. As one might add only `Response(BadRequest)`, the API documentation might as well be plain wrong.

_Example of resource where ErrorMedia and BadRequest would not be reflected in API documentation:_

```
var _ = Resource("bottle", func() {
    BasePath("/bottles")

    Action("create", func() {
        Routing(POST("/"))

        Payload(BottlePayload)

        Response(Created)
    })
})
```

From the error-handling section in the [docs](https://goa.design/implement/error_handling/)

> Note that the controller actions are responsible for implementing the contract defined in the design. That is they should not define error classes that use HTTP status codes not listed in the action definitions.

That _API documentation is a first-class citizen_ and always in sync with the implemented API design is an **absolute stellar characteristic** for an API framework. Unfortunately, this characteristic is currently compromised. I would suspect other default error behavior to currently break this characteristic as well.

I'd be happy to help out if it is reckoned viable with current data models etc to amend (only picked up goa today, not too privy with the code yet).

**Comment by raphael** [comment]
Any validation specified in the design is reflected both in code and in Swagger. If I understand correctly you are alluding to the fact that the design must list the BadRequest response explicitly if it uses input types (payloads and/or params) that define validations. This is true but I hardly see this as a critical issue. BadRequest is the only response that code validation may generate. It would be simple to have the code generation automatically define it but you would then sacrifice the explicitness of the current approach which I find  to be a more important property. This also makes it possible to tweak the response produce by the API in case of a validation issue. While BadRequest with a response media type set to ErrorMedia is the default, that can be easily changed by overriding the corresponding error responses in the goa package. In this case you need to be able to define the BadRequest response yourself to properly describe it.

TL;DR you are responsible for adding the validation violation response so you can tweak it and so that it's clear when reading the design that this response exists.

**Comment by raphael** [comment]
An interesting approach here might be to have the code generation tool emit an error if there are validations defined on an input type but no error response defined. Although the challenge would be to identify that response since it can be overwritten. At three very least there could be an error if only there successful response is defined.

**Comment by frodebjerke** [comment]
I totally buy the point for keeping things explicit! Using the code generation tool to enforce the correct responses definitely seems more interesting.

Pushed a quick naive version of checking for appropriate responses when an action has a payload. It does not address the challenge of identifying the correct MediaType. Additionally, it references behavior in the MuxHandler method without it being coupled. By this, I mean that the `actiondefinition.Validate` method knows MuxHandler generates BadRequest and RequestBodyTooLarge responses. https://github.com/frodebjerke/goa/commit/f0ee0c9f8ea49011d97247d1e9243ff9a4d9ba03

Since the MediaType of errors is handled by a middleware and the middleware chain is not defined in the DSL, identifying the correct MediaType of errors in the code gen tool is challenging/impossible.

Is defining the middleware chain of an API in the DSL a viable option? Are there any good reasons to not define the API in the DSL, except for it being a breaking change to the framework API?

**Comment by raphael** [comment]
I like where this is going :) Just to clarify a few things:

* One may override the validation error response status code by setting the goa package `ErrInvalidRequest` variable - so assuming `BadRequest` is also problematic.
* An action merely having a payload is not a sufficient condition for it to also have validations. The code you have would also have to check that there are validations attached to the payload type.
* There could be an action with no payload that has validations in the params and thus has the same issue of having a generated validation error response. 
* Middlewares don't necessarily affect the design of an API. Some middlewares deal with implementation specific details. I see middlewares as orthogonal when it comes to design vs. implementation. 

That being said it would probably make sense to design validation error responses specifically like you are suggesting. I could see a DSL that would let the designer describe the validation error response. The code generation tool could then use that description to initialize the goa `ErrInvalidRequest` variable on start when present . That definition would have to be in the API DSL, something like:

```go
var _ = API("cellar", func() {
        Description("...")
        ValidationErrorResponse(BadRequest, MyErrorMedia) // Same DSL as `Response`
        // ...
})
```

Not specifying `ValidationErrorResponse` would be equivalent to:

```go
        ValidationErrorResponse(BadRequest, ErrorMedia) // Uses built-in ErrorMedia type
```

I guess this ends up being similar to what you suggested initially with the difference that the specification of the validation error response would be explicit rather than inferred.
